sdk/src/ed25519_instruction.rs

```diff
@@ -0,0 +1,287 @@
+pub enum Ed25519Error {
+    InvalidPublicKey,
+    InvalidSignature,
+    InvalidDataOffsets,
+    InvalidInstructionDataSize,
+}
+
+pub const PUBKEY_SERIALIZED_SIZE: usize = 32;
+pub const SIGNATURE_SERIALIZED_SIZE: usize = 64;
+pub const SIGNATURE_OFFSETS_SERIALIZED_SIZE: usize = 11;
+pub const DATA_START: usize = SIGNATURE_OFFSETS_SERIALIZED_SIZE + 1;
+
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct Ed25519SignatureOffsets {
+    signature_offset: u16,            // offset to ed25519 signature of 64 bytes
+    signature_instruction_index: u8,  // instruction index to find signature
+    public_key_offset: u16,           // offset to public key of 32 bytes
+    public_key_instruction_index: u8, // instruction index to find public key
+    message_data_offset: u16,         // offset to start of message data
+    message_data_size: u16,           // size of message data
+    message_instruction_index: u8,    // index of instruction data to get message data
+}
+
+pub fn new_ed25519_instruction(keypair: &ed25519_dalek::Keypair, message: &[u8]) -> Instruction {
+    let signature = keypair.sign(message).to_bytes();
+    let pubkey = keypair.public.to_bytes();
+
+    assert_eq!(pubkey.len(), PUBKEY_SERIALIZED_SIZE);
+    assert_eq!(signature.len(), SIGNATURE_SERIALIZED_SIZE);
+
+    let mut instruction_data = vec![];
+    instruction_data.resize(
+        DATA_START
+            .saturating_add(signature.len())
+            .saturating_add(pubkey.len())
+            .saturating_add(message.len())
+            .saturating_add(1),
+        0,
+    );
+
+    let public_key_offset = DATA_START;
+    instruction_data[public_key_offset..public_key_offset.saturating_add(pubkey.len())]
+        .copy_from_slice(&pubkey);
+
+    let signature_offset = DATA_START.saturating_add(pubkey.len());
+    instruction_data[signature_offset..signature_offset.saturating_add(signature.len())]
+        .copy_from_slice(&signature);
+
+    let message_data_offset = signature_offset
+        .saturating_add(signature.len())
+        .saturating_add(1);
+    instruction_data[message_data_offset..].copy_from_slice(message);
+
+    let num_signatures = 1;
+    instruction_data[0] = num_signatures;
+    let offsets = Ed25519SignatureOffsets {
+        signature_offset: signature_offset as u16,
+        signature_instruction_index: 0,
+        public_key_offset: public_key_offset as u16,
+        public_key_instruction_index: 0,
+        message_data_offset: message_data_offset as u16,
+        message_data_size: message.len() as u16,
+        message_instruction_index: 0,
+    };
+    let writer = std::io::Cursor::new(&mut instruction_data[1..DATA_START]);
+    bincode::serialize_into(writer, &offsets).unwrap();
+
+    Instruction {
+        program_id: solana_sdk::ed25519_program::id(),
+        accounts: vec![],
+        data: instruction_data,
+    }
+}
+
+pub fn verify_signatures(data: &[u8], instruction_datas: &[&[u8]]) -> Result<(), Ed25519Error> {
```

Comment:
I'll be experimenting with the [bytemuck](https://crates.io/crates/bytemuck) crate a little tomorrow, but if I'm reading the docs.rs right I think we might be able to use that crate here as well to reduce the copies (maybe not quite zero copy) and drop the bincode